### PR TITLE
shell: add 'exit' builtin command

### PR DESCRIPTION
**Pull request by nektro**
breaking out https://github.com/oven-sh/bun/pull/9578 into multiple PRs in order to better pin down where my CI is going awry

edit: progress on https://github.com/oven-sh/bun/issues/9716

### CODE REVIEW

File: src/shell/interpreter.zig [hunk 1]
```diff
@@ -4231,36 +4220,10 @@ pub const Interpreter = struct {
...
-                .ls => {
-                    cmd.exec.bltn.impl = .{
-                        .ls = Ls{
-                            .bltn = &cmd.exec.bltn,
-                        },
-                    };
+                inline else => |tag| {
+                    cmd.exec.bltn.impl = @unionInit(RealImpl, @tagName(tag), .{
+                        .bltn = &cmd.exec.bltn,
+                    });
```
Comment (gvilums) [inline src/shell/interpreter.zig:4232]: Honestly not sure how I feel about this, it's definitely more concise than what we had previously but also makes it harder to understand what exactly is going on IMO.

File: src/shell/interpreter.zig [hunk 2]
```diff
@@ -8663,6 +8626,84 @@ pub const Interpreter = struct {
...
+                return Maybe(void).success;
+            }
+
+            pub fn next(this: *Exit) void {
+                while (!(this.state == .err or this.state == .done)) {
+                    switch (this.state) {
+                        .waiting_io => return,
+                        .idle, .done, .err => unreachable,
+                    }
+                }
```
Comment (gvilums) [inline src/shell/interpreter.zig]: why is this a loop? the body either returns immediately or causes UB

**Review by gvilums — COMMENTED** [review]
I assume the failing tests are just general flakiness

File: src/shell/interpreter.zig (hunk 2 above)
Comment (gvilums) [reply src/shell/interpreter.zig]:
```suggestion
                if(!(this.state == .err or this.state == .done)) {
                    return;
                }
```

File: src/shell/interpreter.zig (hunk 1 above)
Comment (nektro) [reply src/shell/interpreter.zig:4232]: could you elaborate on this a bit?

File: src/shell/interpreter.zig [hunk 3]
```diff
@@ -8669,6 +8632,84 @@ pub const Interpreter = struct {
...
+                    this.state = .err;
+                    this.next();
+                    return;
+                }
+                this.state = .done;
+                this.next();
+            }
+
+            pub fn deinit(this: *Exit) void {
+                _ = this;
```
Comment (Jarred-Sumner) [inline src/shell/interpreter.zig:8710]: is the memory owned by the parent struct?

File: src/shell/interpreter.zig (hunk 1 above)
Comment (Jarred-Sumner) [reply src/shell/interpreter.zig:4232]: personally, i like it

it also encourages all the commands have the same interface

File: src/shell/interpreter.zig (hunk 3 above)
Comment (nektro) [reply src/shell/interpreter.zig:8710]: appears so checking all the other builtins

**Pull request merged**